xl_auth/user/models.py

```diff
@@ -158,6 +159,43 @@ def __repr__(self):
         return '<PasswordReset({email!r})>'.format(email=self.user.email)
 
 
+class LoginAttempt(SurrogatePK, Model):
+    """A failed login attempt."""
+
+    __tablename__ = 'login_attempts'
+    username = Column(db.String(255), unique=False, nullable=False, index=True)
+    source_ip = Column(db.String(255), unique=False, nullable=False, index=True)
+
+    created_at = Column(db.DateTime, default=datetime.utcnow, nullable=False)
+
+    def __init__(self, username, source_ip):
+        """Create instance."""
+        db.Model.__init__(self, username=username, source_ip=source_ip)
+
+    @staticmethod
+    def too_many_recent_failures_for(username):
+        """Determine if a login attempt is allowed or not."""
+        timeframe = current_app.config['FAILED_LOGIN_TIMEFRAME_MINUTES']
+        from_timestamp = datetime.utcnow() - timedelta(minutes=timeframe)
+        attempts = LoginAttempt.query.filter(LoginAttempt.username == username,
+                                             LoginAttempt.created_at >= from_timestamp).all()
+        if len(attempts) >= current_app.config['FAILED_LOGIN_MAX_ATTEMPTS']:
+            return True
```

Comment:
Do we want to log this as a failed attempt too?

Reply:
No, I don't see why. It would be a 429 status in Nginx monitoring, no good idea filling up the live database with junk from a robot hammering the server with password guesses, apart from what's actually needed for throttling guess-rate!

Reply:
Also, from the application point-of-view, I would argue that it's not a login attempt as in "checking the user credentials is OK/NOK", but rather a 429 refusal to process the login request.

Reply:
Also, from the application point-of-view, I would argue that it's not a login attempt as in "checking the user credentials is OK/NOK", but rather a 429 refusal to process the login request.

Reply:
Good point with the nginx log! I'll keep it as-is.